--- a/arch/arm64/boot/dts/rockchip/rk3399-orangepi-4-lts.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-orangepi-4-lts.dts
@@ -209,16 +209,36 @@
 	assigned-clocks = <&cru SCLK_RMII_SRC>;
 	assigned-clock-parents = <&clkin_gmac>;
 	clock_in_out = "input";
+	pinctrl-names = "default";
+	pinctrl-0 = <&rgmii_pins>, <&phy_rstb>;
+	phy-mode = "rgmii";
 	phy-supply = <&vcc3v3_s3>;
-	phy-mode = "rgmii";
-	pinctrl-names = "default";
-	pinctrl-0 = <&rgmii_pins>;
-	snps,reset-gpio = <&gpio3 RK_PB7 GPIO_ACTIVE_LOW>;
+	phy-handle = <&yt8531c>;
 	snps,reset-active-low;
 	snps,reset-delays-us = <0 10000 50000>;
+	snps,reset-gpio = <&gpio3 RK_PB7 GPIO_ACTIVE_LOW>;
 	tx_delay = <0x28>;
 	rx_delay = <0x11>;
 	status = "okay";
+
+	mdio {
+		compatible = "snps,dwmac-mdio";
+        	#address-cells = <1>;
+	        #size-cells = <0>;
+	        yt8531c: ethernet-phy@0 {
+			compatible = "ethernet-phy-id4f51.e91b",
+				     "ethernet-phy-ieee802.3-c22";
+            		reg = <0>;
+			pinctrl-names = "default";
+			pinctrl-0 = <&phy_intb>;
+            		motorcomm,clk-out-frequency-hz = <125000000>;
+            		motorcomm,keep-pll-enabled;
+            		motorcomm,auto-sleep-disabled;
+			interrupt-parent = <&gpio3>;
+			interrupts = <RK_PB2 IRQ_TYPE_LEVEL_LOW>;
+        	};
+    	};
+
 };
 
 &gpu {

