--- a/arch/riscv/Makefile
+++ b/arch/riscv/Makefile
@@ -127,10 +127,10 @@
 ifeq ($(CONFIG_RISCV_M_MODE)$(CONFIG_SOC_CANAAN),yy)
 KBUILD_IMAGE := $(boot)/loader.bin
 else
-KBUILD_IMAGE := $(boot)/Image.gz
+KBUILD_IMAGE := $(boot)/Image
 endif
 endif
-BOOT_TARGETS := Image Image.gz loader loader.bin xipImage
+BOOT_TARGETS := Image loader loader.bin xipImage
 
 all:	$(notdir $(KBUILD_IMAGE))
 
@@ -142,7 +142,7 @@
 	$(Q)$(MAKE) $(build)=$(boot) $(boot)/$@
 
 install: KBUILD_IMAGE := $(boot)/Image
-zinstall: KBUILD_IMAGE := $(boot)/Image.gz
+zinstall: KBUILD_IMAGE := $(boot)/Image
 install zinstall:
 	$(call cmd,install)
 

