--- a/scripts/package/mkdebian
+++ b/scripts/package/mkdebian
@@ -103,7 +103,6 @@
 packagename=linux-image-"$NAME_EXTENSION"
 kernel_headers_packagename=linux-headers-"$NAME_EXTENSION"
 libc_headers_packagename=linux-libc-dev
-dtb_packagename=linux-dtb-"$NAME_EXTENSION"
 dbg_packagename=$packagename-dbg
 
 if [ "$ARCH" = "um" ] ; then
@@ -204,12 +203,6 @@
  This package provides userspaces headers from the Linux kernel.  These headers
  are used by the installed headers for GNU glibc and other system libraries.
 Multi-Arch: same
-
-Package: $dtb_packagename
-Architecture: $debarch
-Provides: linux-dtb, linux-dtb-$NAME_EXTENSION
-Description: Armbian Linux DTB, version $version $NAME_EXTENSION
- This package contains device blobs from the Linux kernel, version $version
 EOF
 
 if is_enabled CONFIG_MODULES; then

